keps/sig-api-machinery/0015-dry-run.md

```diff
@@ -150,3 +151,52 @@ A dry-run request should behave as close as possible to a regular
 request. Attempting to dry-run create an existing object will result in an
 `AlreadyExists` error to be returned. Similarly, if a dry-run update is
 performed on a non-existing object, a `NotFound` error will be returned.
+
+## kubectl
+
+For the `kubectl` client integration with server-side dry-run, we will pass
+the `dryRun` query parameter by reading the user's intent from a flag.
+
+For beta, we use `--server-dry-run` for `kubectl apply` to exercise
+server-side apply. This flag will be deprecated next release, then removed after 1 release.
+
+For GA, we'll use the existing `--dry-run` flag available on `kubectl apply`.
+
+Currently, the `--dry-run` flag is a boolean for subcommands including
+`kubectl apply` and `kubectl create`.
+
+We'll extend the flag to accept strings for new options
+for selecting client-side and server-side behavior: `client`, `server`, and
+`none`.
+
+For backwards compatibility, we'll continue to default the value for `--dry-run`
+to `--dry-run=client`, which is equivalent to the existing behavior for
```

Comment:
I don't think is the place where you want to end up.  I expected the command to eventually have `--dry-run` mean "server-side dry run".  How would you manage the transition of the default?

Reply:
As a human user or in CI, the primary use case is to validate that `kubectl apply --dry-run -f path/to/manifests` works. We do want to change the default to `--dry-run=server`, and the expectation would not be broken for human users or CI with the new default.

I'll update the KEP to say we're introducing `--dry-run=client` for backward compatibility in this release (1.17), but will deprecate this option in 1 release after (1.18), and remove it in 3 releases (1.20)? Such that we end in a state where `none` and `server` are the only supported values in 3 releases (1.20), with `--dry-run=server` being the default.

Reply:
@apelisse mentioned you discussed ensuring the dry run flag is always explicitly set instead of implicitly defaulted.

Updated the KEP to deprecate the default unset flag behavior and require that the flag is always set in the future.

PTAL @deads2k @soltysh